superset/reports/commands/execute.py

```diff
@@ -0,0 +1,256 @@
+    - On reports takes chart or dashboard screenshots and sends configured notifications
+    - On Alerts uses related Command AlertCommand and sends configured notifications
+    """
+
+    def __init__(self, model_id: int, scheduled_dttm: datetime):
+        self._model_id = model_id
+        self._model: Optional[ReportSchedule] = None
+        self._scheduled_dttm = scheduled_dttm
+
+    def set_state_and_log(
+        self,
+        session: Session,
+        start_dttm: datetime,
+        state: ReportLogState,
+        error_message: Optional[str] = None,
+    ) -> None:
+        """
+        Updates current ReportSchedule state and TS. If on final state writes the log
+        for this execution
+        """
+        now_dttm = datetime.utcnow()
+        if state == ReportLogState.WORKING:
+            self.set_state(session, state, now_dttm)
+            return
+        self.set_state(session, state, now_dttm)
+        self.create_log(
+            session, start_dttm, now_dttm, state, error_message=error_message,
+        )
+
+    def set_state(
+        self, session: Session, state: ReportLogState, dttm: datetime
+    ) -> None:
+        """
+        Set the current report schedule state, on this case we want to
+        commit immediately
+        """
+        if self._model:
+            self._model.last_state = state
+            self._model.last_eval_dttm = dttm
+            session.commit()
+
+    def create_log(  # pylint: disable=too-many-arguments
+        self,
+        session: Session,
+        start_dttm: datetime,
+        end_dttm: datetime,
+        state: ReportLogState,
+        error_message: Optional[str] = None,
+    ) -> None:
+        """
+        Creates a Report execution log, uses the current computed last_value for Alerts
+        """
+        if self._model:
+            log = ReportExecutionLog(
+                scheduled_dttm=self._scheduled_dttm,
+                start_dttm=start_dttm,
+                end_dttm=end_dttm,
+                value=self._model.last_value,
+                value_row_json=self._model.last_value_row_json,
+                state=state,
+                error_message=error_message,
+                report_schedule=self._model,
+            )
+            session.add(log)
+
+    @staticmethod
+    def _get_url(report_schedule: ReportSchedule, user_friendly: bool = False) -> str:
+        """
+        Get the url for this report schedule: chart or dashboard
+        """
+        if report_schedule.chart:
+            return get_url_path(
+                "Superset.slice",
+                user_friendly=user_friendly,
+                slice_id=report_schedule.chart_id,
+                standalone="true",
+            )
+        return get_url_path(
+            "Superset.dashboard",
+            user_friendly=user_friendly,
+            dashboard_id_or_slug=report_schedule.dashboard_id,
+        )
+
+    def _get_screenshot(self, report_schedule: ReportSchedule) -> ScreenshotData:
```

Comment:
nit: it would be nice to pass only the attributes needed to calculate the screenshot vs a whole class
it makes it easier to unit tests the functions when less objects needs to be constructed, and read the functions / code as well as functions definition explains what is passed to it. 

this is just a personal preference and optional suggestion

Reply:
Good point, yet commands internally are model centric, we assume that `validate` validates and populates the model by `model_id`. My take on unit tests is to test public methods, and assume that the screenshot itself is being tested also